plugin.video.liveleak/default.py

```diff
@@ -1,196 +1,311 @@
-               'Accept-Language':'en-US,en;q=0.8'
-                } 
-
-    req = urllib2.Request(url.encode('utf-8'), None, headers)
-    if method is 'head': req.get_method = lambda : 'HEAD'
-    try:
-        response = urllib2.urlopen(req)
-        if method is 'head':
-            text = response.info()
+def buildUrl(query):
+    return BASE_URL + '?' + urlencode(query)
+
+def findAllMediaItems(block):
+    media = []
+    # Consolidate liveleak and Youtube video sources
+    liveleak_vids = block.select('video > source')
+    for vid in liveleak_vids:
+        media.append(vid['src'])
+
+    youtubes = []
+    youtube_vids = block.select('iframe[src*="youtube.com/embed/"]')
+    for vid in youtube_vids:
+        # Extract youtube video id and append to list
+        youtubes.append(re.search(r'youtube.com/embed/(.+?)\?rel=0.*$', vid['src']).group(1))
+    youtubes = list(set(youtubes)) # remove duplicates
+
+    media.extend(youtubes) #concatenate liveleak & youtube media references
+
+    return media
+
+def fetchItemDetails((url, meta)):
+    page = requests.get(url)
+
+    if page.status_code != requests.codes.ok: # Uh-oh!
+        log("Error: %s: %s" % (page.status_code, url))
+        return
+
+    # Reduce page to video information block
+    soup = bs(page.text, 'html.parser')
+    vid_block = soup.find('div', class_='step_outer')
+    for script in vid_block("script"):
+        script.decompose()
+    for style in vid_block("style"):
+        style.decompose()
+
+    # Get post description
+    meta['description'] = vid_block.get_text().strip()
+
+    media = findAllMediaItems(vid_block)
+    if media:
+        if len(media) > 1: # More than one media item on page
+            mediaList = []
+            for medium in media:
+                mediaList.append((url, medium, meta))
+            return mediaList
         else:
-            text = response.read()
-        response.close()
-    except:
-        text = None
+            return ((url, media[0], meta)) # single item
+
+def buildListItem((url, medium, meta)):
+    #Extract meta info
+    title = meta['title']
+    thumbnail = meta['thumbnail']
+    credit = meta['credit']
+    description = meta['description']
+    mpaa = meta['mpaa']
+    rating = meta['rating']
+
+    leakPosters = loadLeakPosters() # Preferences for coloring titles
+
+    if 'cdn.liveleak.com' in medium:
+        # Capture source of this medium
+        src = url.replace(domain_home, '')
+        url = buildUrl({'mode': 'play', 'url': medium, 'src': src})
+    else:
+        url = 'plugin://plugin.video.youtube/play/?video_id=%s' % medium
+        url = url.encode('utf-8')
+
+    # Build list item
+    user_mod = leakPosters.get(credit, 0)
+    if user_mod == 1:
+        title = "[COLOR limegreen]%s[/COLOR]" % title
+    elif user_mod == 2:
+        title = "[COLOR dimgray]%s[/COLOR]" % title
+    liz = xbmcgui.ListItem(label=title, thumbnailImage=thumbnail)
```

Comment:
It would be nice if you could remove the `thumbnailImage` kwarg, it is deprecated in favor of `listitem.setArt`